aten/src/ATen/native/cuda/TensorModeKernel.cu

```diff
@@ -0,0 +1,322 @@
+  // Ideally we would have one handle_fused_mode for each power of 2
+  switch (ceilPowerOf2) {
+    case 2048:
+      handle_fused_mode<2048, scalar_t>(
+          grid, self, ti_values, ti_indices, slice_size);
+      break;
+    case 1024:
+    case 512:
+    case 256:
+      handle_fused_mode<1024, scalar_t>(
+          grid, self, ti_values, ti_indices, slice_size);
+      break;
+    case 128:
+    case 64:
+      handle_fused_mode<128, scalar_t>(
+          grid, self, ti_values, ti_indices, slice_size);
+      break;
+    case 32:
+    case 16:
+    case 8:
+    case 4:
+    case 2:
+      handle_fused_mode<32, scalar_t>(
+          grid, self, ti_values, ti_indices, slice_size);
+      break;
+    case 1:
+    default:
+      TORCH_INTERNAL_ASSERT(false);
+  }
+
+  AT_CUDA_CHECK(cudaGetLastError());
+}
+
+void mode_kernel_impl(
+    Tensor& values,
+    Tensor& indices,
+    const Tensor& self,
+    int64_t dim,
+    bool keepdim) {
+  auto self_sizes = ensure_nonempty_vec(self.sizes().vec());
+  int64_t ndim = ensure_nonempty_dim(self.dim());
+  int64_t slice_size = ensure_nonempty_size(self, dim);
+  int64_t slices = self.numel() / slice_size;
+
+  // Resize output value, index Tensors to appropriate sizes (i.e. the same as
+  // the input Tensor, except at dim=dimension, the size is 1)
+  self_sizes[dim] = 1;
+
+  if (!keepdim) {
+    if (values.ndimension() >= dim) {
+      values.unsqueeze_(dim);
+    }
+    if (indices.ndimension() >= dim) {
+      indices.unsqueeze_(dim);
+    }
+  }
+
+  values.resize_(self_sizes);
+  indices.resize_(self_sizes);
+
+  // If sliceSize is 1, copy input to values and set indices
+  if (slice_size == 1) {
+    values.copy_(self);
+    indices.fill_(0);
+    if (!keepdim) {
+      values.squeeze_(dim);
+      indices.squeeze_(dim);
+    }
+    return;
+  }
+
+  // Beginning our optimized implementation. First thing we want to do is to
+  // transpose the input Tensor along the sort dimension, and then make it
+  // contiguous.
+  auto transposed = self.transpose(dim, ndim - 1);
+  auto contiguous = transposed.contiguous();
+
+  // We also need to view the values and indices Tensors as transposed in order
+  // to properly determine the offset into the underlying storage in which to
+  // place the mode and index for a particular set of dimension values.
+  auto values_transposed = values.transpose(dim, ndim - 1);
+  auto indices_transposed = indices.transpose(dim, ndim - 1);
+
+  // Call mode
+  AT_DISPATCH_ALL_TYPES_AND(kHalf, self.scalar_type(), "cuda_mode", [&] {
+    // Requirements for fused kernel implementation:
+    //
+    // 1. sliceSize <= 2 * max threads per block
+    // 2. uses one block per slice, so number of slices must be less than the
+    // maximum number of blocks for a kernel launch
+    // 3. Can use 32-bit index math for indexing (mainly just for implementation
+    // conciseness, could be changed)
+    if (slice_size <= MAX_BLOCK_SIZE && slices <= MAX_GRID_SIZE &&
```

Comment:
slices <= MAX_GRID_SIZE looks too restrictive, `getGridFromTiles` requires slices <= MAX_GRID_SIZE * MAX_GRID_SIZE * MAX_GRID_SIZE`. Relaxing this would probably fix #46507.
Can you also add a test for > 65535 slices?

Reply:
Ah, I hadn't noticed that. That makes sense, indeed.
I believe it makes sense to rewrite it as `slices <= MAX_GRID_SIZE * MAX_GRID_SIZE * MAX_GRID_SIZE`, right!?

Reply:
Yes, that should be the correct condition